### PR TITLE
fix(ui): Fix absolute date picker

### PR DESCRIPTION
**Pull request by billyvg**
Fixes absolute date picker when system timezone differs from account timezone.

### CODE REVIEW

File: src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/utils.jsx [hunk 1]
```diff
@@ -26,8 +28,8 @@ export function getStateFromQuery(query) {
   }
 
   if (hasAbsolute) {
-    start = getLocalDateObject(start);
-    end = getLocalDateObject(end);
+    start = getUtcToLocalDateObject(start);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/globalSelectionHeader/utils.jsx]: This is just renamed to maybe be more clear.

File: src/sentry/static/sentry/app/components/organizations/timeRangeSelector/dateRange/dateRange.jsx [hunk 2]
```diff
@@ -74,29 +68,20 @@ class DateRange extends React.Component {
     maxPickableDays: MAX_PICKABLE_DAYS,
   };
 
-  static getTimeStringFromDate = (date, utc) => {
-    return getFormattedDate(date, 'HH:mm', {local: !utc});
+  static getTimeStringFromDate = date => {
+    return moment(date)
+      .local()
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/timeRangeSelector/dateRange/dateRange.jsx]: Date picker only deals with system timezone.

File: src/sentry/static/sentry/app/components/organizations/timeRangeSelector/dateRange/dateRange.jsx [hunk 3]
```diff
@@ -74,29 +68,20 @@ class DateRange extends React.Component {
...
-    if (start) {
-      start = setDateToTime(start, DEFAULT_DAY_START_TIME, {local: !utc});
-    }
-
-    if (end) {
-      end = setDateToTime(end, DEFAULT_DAY_END_TIME, {
-        local: !utc,
-      });
-    }
+    const end = endDate ? getEndOfDay(endDate) : endDate;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/organizations/timeRangeSelector/dateRange/dateRange.jsx]: When selecting a date range, the start/end times are at 00:00, to make end date inclusive, need to get the end of the day.

File: src/sentry/static/sentry/app/stores/globalSelectionStore.jsx [hunk 4]
```diff
@@ -119,10 +119,7 @@ const GlobalSelectionStore = Reflux.createStore({
 
     this.selection = {
       ...this.selection,
-      datetime: {
-        ...this.selection.datetime,
-        ...datetime,
-      },
+      datetime,
```
Comment (billyvg) [inline src/sentry/static/sentry/app/stores/globalSelectionStore.jsx]: There was a slight bug when switching from relative to absolute would preserve the relative value before then being updated again with the correct `relative: null`.

File: src/sentry/static/sentry/app/views/organizationDiscover/utils.jsx [hunk 5]
```diff
@@ -18,7 +18,7 @@ const VALID_QUERY_KEYS = [
 ];
 
 export function getQueryFromQueryString(queryString) {
-  const queryKeys = new Set(VALID_QUERY_KEYS);
+  const queryKeys = new Set([...VALID_QUERY_KEYS, 'utc']);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationDiscover/utils.jsx]: Need to preserve `utc` state in url

**Comment by billyvg** [comment]
Testing:
- Tested with account timezone set to something different than system timezone AND with account == system
- Tested on (Events|Releases|Issues) AND Discover
- Time Inputs
- Switching from Relative -> Absolute
- Switching UTC override

Known issues:
- Relative --> Absolute --> UTC (without closing menu in between)

File: src/sentry/static/sentry/app/components/organizations/timeRangeSelector/index.jsx [hunk 6]
```diff
@@ -73,21 +97,48 @@ class TimeRangeSelector extends React.PureComponent {
...
+    if (!datetime.start && !datetime.end) {
+      callback(datetime);
+      return;
+    }
+
+    // Change local date into either UTC or local time (local time defined by user preference)
+    callback({
+      ...datetime,
+      start: getDateWithTimezoneInUtc(datetime.start, this.state.utc),
+      end: getDateWithTimezoneInUtc(datetime.end, this.state.utc),
```
[the remark this answers is not in the document]
Comment (billyvg) [reply src/sentry/static/sentry/app/components/organizations/timeRangeSelector/index.jsx]: `datetime` includes the current `utc` state

**Review by lynnagara — APPROVED** [review]
This is complicated and a little tricky to follow but mostly looks good to me 🙂

**Pull request merged**